File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/integration/booking/handlers/CreditFromBooking.java
package it.giannibombelli.wsc2026.giftcard.application.integration.booking.handlers;

import it.giannibombelli.wsc2026.booking.integration.giftcard.BookingResultIntegrationEvent;
import it.giannibombelli.wsc2026.giftcard.application.commands.CreditGiftCard;
import it.giannibombelli.wsc2026.giftcard.application.integration.booking.adapter.BookingResult;
import it.giannibombelli.wsc2026.giftcard.application.usecases.GiftCardCrediting;

import static java.util.Objects.requireNonNull;

public final class CreditFromBooking {
    private final BookingResult bookingResult;
    private final GiftCardCrediting useCase;

    public CreditFromBooking(BookingResult bookingResult, GiftCardCrediting giftCardCrediting) {
        this.bookingResult = requireNonNull(bookingResult);
        this.useCase = requireNonNull(giftCardCrediting);
    }

    public void handle(BookingResultIntegrationEvent event) {
        final CreditGiftCard cmd = bookingResult.adapt(event);

        if (cmd != null) {
            useCase.invoke(cmd);
        }
    }
}

File: typescript/src/giftcard/application/integration/booking/handlers/creditFromBooking.ts
import type { BookingResultIntegrationEvent } from '@/booking/integration/giftcard/bookingResultIntegrationEvent.js';
import { BookingResult } from '@/giftcard/application/integration/booking/adapter/bookingResult.js';
import { GiftCardCrediting } from '@/giftcard/application/usecases/giftCardCrediting.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class CreditFromBooking {
  private readonly _bookingResult: BookingResult;
  private readonly _useCase: GiftCardCrediting;

  constructor(bookingResult: BookingResult, giftCardCrediting: GiftCardCrediting) {
    requireDependency(bookingResult, 'bookingResult');
    this._bookingResult = bookingResult;
    requireDependency(giftCardCrediting, 'giftCardCrediting');
    this._useCase = giftCardCrediting;
  }

  handle(event: BookingResultIntegrationEvent): void {
    const command = this._bookingResult.adapt(event);

    if (command !== null) {
      this._useCase.invoke(command);
    }
  }
}
